habitat_code/kobylarz_kmeans_review_code_20240313.py

```diff
@@ -0,0 +1,603 @@
+            result_attr = result['umm']
+            result_gran_id = result_attr['GranuleUR']
+            if result_gran_id == gran_id:
+                extent = result_attr['SpatialExtent']
+                geom_extent=extent['HorizontalSpatialDomain']['Geometry']['GPolygons']
+                coord_list = []
+                for item in geom_extent:
+                    poly = item['Boundary']['Points']
+                    for vertex in poly:
+                        longitude = vertex['Longitude']
+                        latitude = vertex['Latitude']
+                        coord_list.append((longitude, latitude))
+                granule_geometry = shapely.geometry.Polygon(coord_list)
+        # Append info for each file to dataframe
+        new_row = pd.DataFrame({'geometry': [granule_geometry],
+                   'granule_id': [gran_id],
+                   'file_url': [file_url],
+                   'tile_id': [tile_id],
+                   'datetime': [datetime],
+                   'band_number': [band]
+                  })
+        granule_df = pd.concat([granule_df, new_row], ignore_index=True)
+    return granule_df
+
+
+# In[8]:
+
+
+@cached("download_df", False)
+def do_something(results, open_results):
+    download_df = create_granule_df(results, open_results)
+    return download_df
+download_df = do_something(results, open_results)
+
+
+# ### Open, crop, and mask data
+# 
+# This will be the most resource-intensive step. I recommend caching your results using the `cached` decorator or by writing your own caching code. I also recommend testing this step with one or two dates before running the full computation.
+# 
+# This code should include at least one **function** including a numpy-style docstring. A good place to start would be a function for opening a single masked raster, applying the appropriate scale parameter, and cropping.
+# 
+# **YOUR TASK:**
+# 1. For each granule:
+#     1. Open the Fmask band, crop, and compute a quality mask for the granule. You can use the following code as a starting point, making sure that `mask_bits` contains the quality bits you want to consider:
+#          ```python
+#          # Expand into a new dimension of binary bits
+#          bits = np.unpackbits(da.astype(np.uint8)).reshape(da.shape + (-1,))
+# 
+#          # Select the required bits and check if any are flagged
+#          mask = np.prod(bits[..., mask_bits]==0, axis=-1)
+#          ```
+# 
+#     2. For each band that starts with 'B':
+#         1. Open the band, crop, and apply the scale factor
+#         2. Name the DataArray after the band using the `.name` attribute
+#         3. Apply the cloud mask using the `.where()` method
+#         4. Store the DataArray in your data structure (e.g. adding a GeoDataFrame column with the DataArray in it. Note that you will need to remove the rows for unused bands)
+
+# In[9]:
+
+
+# Function to compute image mask
+def process_tifs(download_df, huc_gdf):
```

Comment:
What might have made this more modular is to write this function for only a single granule and then loop it to get your arrays.   That way someone running this on their own could more easily specify if they only wanted a few specific granules for inspection / debugging without having to significantly rewrite the code.